### ISSUE TITLE
inference depth with same results

### ISSUE THREAD

**Issue by liumingcun** [ClementPinard/SfmLearner-Pytorch#125]
Hello author, I have a problem. No matter what picture I use, the final depth image and disp image are exactly the same. I don’t know the reason. Could you please help me solve it? Thank you very much .
![0000000160](https://user-images.githubusercontent.com/23336130/121535478-cbb9e580-ca34-11eb-8e20-59006318a3ed.png)
![0000000000](https://user-images.githubusercontent.com/23336130/121535488-ce1c3f80-ca34-11eb-99dc-aa319662f9b5.png)
![0000000000_depth](https://user-images.githubusercontent.com/23336130/121535589-e0967900-ca34-11eb-8e8d-6c14a656360d.png)
![0000000000_disp](https://user-images.githubusercontent.com/23336130/121535593-e1c7a600-ca34-11eb-800c-9a1829596ee3.png)
![0000000001_depth](https://user-images.githubusercontent.com/23336130/121535595-e2603c80-ca34-11eb-91eb-f7d4977573fa.png)
![0000000001_disp](https://user-images.githubusercontent.com/23336130/121535597-e2603c80-ca34-11eb-8bc6-a699d9e896b8.png)

**Comment by ClementPinard** [comment]
Hi, the pretrained network is especially trained to work on KITTI images that were resized. As such, it will only work with images that are 416x128that share the same intrinsics as KITTI.

What you can do though is try to retrain the network with videos from your own camera, and then pretrained network will work for your image.

Bear in mind that the network is very specialized so if during thraining it only sees road pictures (as it is the cas in KITTI), the quality will be poor with different scenes

**Comment by liumingcun** [comment]
Hello author, thank you very much for your answer. However, I did use kitti data when I tested, but the results are exactly the same as the results I showed. 
![image](https://user-images.githubusercontent.com/23336130/121617985-95ae4d00-ca98-11eb-94f9-f21556a549b0.png)
![image](https://user-images.githubusercontent.com/23336130/121617945-8202e680-ca98-11eb-8f9d-2ddf87428ee6.png)

**Comment by ClementPinard** [comment]
- Is the network the pretrained you can download in the readme or did you train it yourself ? Some hyperparameters (especially a too strong smooth loss) will get you this degenerate network that always output the same thing.
- Are the images resized to 416x128 ?
- Are you feeding DispNet normalize pixel values, i.i. colors are not in [0, 255] but in [-1, 1] ?

**Comment by liumingcun** [comment]
Thank you author. I figured it out. The reason is I feed Dispnet normalize pixel values.

**Comment by Rashfu** [comment]
> Thank you author. I figured it out. The reason is I feed Dispnet normalize pixel values.

Hi, liumingcun
I met the same question as you. We should feed Dispnet pixel values in [0, 255] ? I wonder if I need to modify the code in run_inference.py

**Comment by Rashfu** [comment]
> * Is the network the pretrained you can download in the readme or did you train it yourself ? Some hyperparameters (especially a too strong smooth loss) will get you this degenerate network that always output the same thing.
> * Are the images resized to 416x128 ?
> * Are you feeding DispNet normalize pixel values, i.i. colors are not in [0, 255] but in [-1, 1] ?

Hi, @ClementPinard :
I download pretrained model from the link in README，and when I test on KITTI images, I just use run_inference.py(without any modification) to get the final depth image and disp image which are exactly the same as below. 
![0000000000](https://user-images.githubusercontent.com/42403175/122770099-b2882300-d2d7-11eb-84b7-44aa46b657e9.png)
![0000000000_depth](https://user-images.githubusercontent.com/42403175/122770117-b74cd700-d2d7-11eb-883b-8ff06aa7aec2.png)
![0000000010_disp](https://user-images.githubusercontent.com/42403175/122770133-b9169a80-d2d7-11eb-9109-3c50ad5a1109.png)

As you mentioned before: 

> Are you feeding DispNet normalize pixel values, i.i. colors are not in [0, 255] but in [-1, 1]

But I didn't change any code, just load some images. How could I fix it ? Thanks a lot~

**Comment by ClementPinard** [comment]
You picture proposition lacks a bit of structure. Can you try with a more urban scene ? I am suprised by the results as it looks very blurry but at the same time, you are inside of a vegetation tunnel for the lower part of the screen :joy: 

What do you mean by "I didn't change any code, just load some image" what were the arguments you used in the inference_script ?

**Comment by Rashfu** [comment]
> You picture proposition lacks a bit of structure. Can you try with a more urban scene ? I am surprised by the results as it looks very blurry but at the same time, you are inside of a vegetation tunnel for the lower part of the screen 😂
> 
> What do you mean by "I didn't change any code, just load some image" what were the arguments you used in the inference_script ?

I change some more urban scene but the results looks still the same.
![0000000098](https://user-images.githubusercontent.com/42403175/122848257-c1a0bc80-d33b-11eb-8db4-580276543ecc.png)
![0000000098_depth](https://user-images.githubusercontent.com/42403175/122848263-c5344380-d33b-11eb-9f5e-dabf80b63528.png)
![0000000098_disp](https://user-images.githubusercontent.com/42403175/122848269-c6fe0700-d33b-11eb-8856-17d39607d196.png)

I save some urban scene imgs in `./imgs`, pretrained model in `./weights`. Then I seperately run the `run_inference.py` as below:
```
python3 run_inference.py --pretrained ./weights/dispnet_model_best.pth.tar --dataset-dir ./imgs --output-dir ./imgs_out/disp --output-disp
python3 run_inference.py --pretrained ./weights/dispnet_model_best.pth.tar --dataset-dir ./imgs --output-dir ./imgs_out/depth --output-depth
```
I just feel it strange to generate invariant depth and disparity. 😔 Any mistakes I have made? 😟

**Comment by ClementPinard** [comment]
Hey, sorry all about everything, there is a bug in the inference_script . Indeed, I changed the image loading function to use `imageio.imread` and didn't test thoroughly enough. As a consequence I didn't see that output of `imageio.imread` was in [0,1] and not [0, 255].

I changed the Line 66 from 
```python
        tensor_img = ((tensor_img/255 - 0.5)/0.5).to(device)
```
to
```python
        tensor_img = ((tensor_img - 0.5)/0.5).to(device)
```

And it works much better. A fix is coming, but you can already directly applied the change I mentionned to you.

Clément

**Comment by Rashfu** [comment]
> Hey, sorry all about everything, there is a bug in the inference_script . Indeed, I changed the image loading function to use `imageio.imread` and didn't test thoroughly enough. As a consequence I didn't see that output of `imageio.imread` was in [0,1] and not [0, 255].
> 
> I changed the Line 66 from
> 
> ```python
>         tensor_img = ((tensor_img/255 - 0.5)/0.5).to(device)
> ```
> 
> to
> 
> ```python
>         tensor_img = ((tensor_img - 0.5)/0.5).to(device)
> ```
> 
> And it works much better. A fix is coming, but you can already directly applied the change I mentionned to you.
> 
> Clément

That works fine! Thanks a lot!

**Comment by liumingcun** [comment]
> > Hey, sorry all about everything, there is a bug in the inference_script . Indeed, I changed the image loading function to use `imageio.imread` and didn't test thoroughly enough. As a consequence I didn't see that output of `imageio.imread` was in [0,1] and not [0, 255].
> > I changed the Line 66 from
> > ```python
> >         tensor_img = ((tensor_img/255 - 0.5)/0.5).to(device)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to
> > ```python
> >         tensor_img = ((tensor_img - 0.5)/0.5).to(device)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And it works much better. A fix is coming, but you can already directly applied the change I mentionned to you.
> > Clément
> 
> That works fine! Thanks a lot!

Hello, my WeChat is 879997125, I think we can exchange related questions

**Comment by liumingcun** [comment]
> > Thank you author. I figured it out. The reason is I feed Dispnet normalize pixel values.
> 
> Hi, @liumingcun:
> I met the same question as you. Should we feed Dispnet pixel values in [0, 255] ? I haven't read the code carefully, so I wonder if there is a need to modify the code in run_inference.py

Hello, my WeChat is 879997125, I think we can exchange related questions

**Comment by Rashfu** [comment]
> > > Thank you author. I figured it out. The reason is I feed Dispnet normalize pixel values.
> > 
> > 
> > Hi, @liumingcun:
> > I met the same question as you. Should we feed Dispnet pixel values in [0, 255] ? I haven't read the code carefully, so I wonder if there is a need to modify the code in run_inference.py
> 
> Hello, my WeChat is 879997125, I think we can exchange related questions

Yeah ,that's good idea~

**Comment by Rashfu** [comment]
> > Hey, sorry all about everything, there is a bug in the inference_script . Indeed, I changed the image loading function to use `imageio.imread` and didn't test thoroughly enough. As a consequence I didn't see that output of `imageio.imread` was in [0,1] and not [0, 255].
> > I changed the Line 66 from
> > ```python
> >         tensor_img = ((tensor_img/255 - 0.5)/0.5).to(device)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to
> > ```python
> >         tensor_img = ((tensor_img - 0.5)/0.5).to(device)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And it works much better. A fix is coming, but you can already directly applied the change I mentionned to you.
> > Clément
> 
> That works fine! Thanks a lot!

Hi, @ClementPinard :
The output of `imageio.imread` was still in [0, 255]. The bug is in `skimage.transform.resize`: if the `dtype` of input img is `uint8` then the `resize` function will normalize the value in [0, 1] automatically, while in`float32` the `resize` function just do the resize.
